### PR TITLE
[ROCm] Fix layer_norm fallback launch limit

### LINKED ISSUE #184826: ROCm `F.layer_norm` silently produces wrong output for `M >= 2^23, C%4!=0`
### 🐛 Describe the bug

This debugging and this issue report were largely created by Claude Code. I haven't checked its exact process, but the repro I think demonstrates it pretty clearly.

On ROCm, `torch.nn.functional.layer_norm` silently produces incorrect output (zeros or what looks like uninitialized memory) for the back portion of large 2D-reshaped inputs. Two related fixes have already landed for this same kernel (#144007, #181600), but neither resolves the failures shown below.

The bug is deterministic with `torch.manual_seed(0)`, reproduces on two different AMD GPU architectures, and reproduces on the current PyTorch nightly that already contains the #181600 fix.

## Motivating Shapes

`(1, 1, 2048, 5247, 267) M*N=2,869,143,552 (0.67x 2^32)`
`(1, 1, 5247, 5247, 267) M*N=7,350,779,403 (1.71x 2^32)`

| GPU                           | Arch    | PyTorch                        | Shape A (\<2^32) | Shape B (>2^32)             |
| ----------------------------- | ------- | ------------------------------ | ---------------- | --------------------------- |
| AMD Instinct MI300X           | gfx942  | `2.10.0+rocm7.2.0.gitb6ee5fde` | bad              | bad                         |
| AMD Instinct MI300X           | gfx942  | `2.13.0.dev20260521+rocm7.2`   | bad              | bad                         |
| AMD Radeon 8060S (Strix Halo) | gfx1151 | `2.11.0+rocm7.2`               | bad              | (not tested — OOM at 60 GB) |

The nightly `2.13.0.dev20260521+rocm7.2` was built ~17 days after #181600 was merged (commit 8eae40c5), and the fix is present in its sources, but both shapes still fail.

The failure pattern is identical across all three configurations: bad row count is a multiple of 1599, bad rows always end at `d2-1`, first bad row is 449 or 450. Independent of dtype (fp32 and bf16 both fail at the same shape).

## Specific features

- CPU `F.layer_norm` on the same saved input produces correct output

- A chunked GPU `F.layer_norm` (slice-by-slice along a leading dim) on the same input produces correct output, matching CPU.

- Bad rows in the output are either zero or filled with values that look like uninitialized memory — e.g. consecutive entries differing by exactly 16384:

  ```text
  [101877579776.0, 101877596160.0, 101877612544.0, 101877628928.0, ...]
  ```

- When M == 2^23, error "invalid configuration argument"

- When M > 2^23 and C % 4 != 0 silently wrong output

<details>
<summary>Minimal reproduction</summary>

```python
import math

import torch
import torch.nn.functional as F


@torch.no_grad()
def check(M, C, dtype=torch.bfloat16, atol=0.05, chunk=8192):
    """Run F.layer_norm and compare against a row-chunked reference.
    Returns bad-row count, first-bad-row index, and NaN-row count."""
    torch.manual_seed(0)
    x = torch.randn(M, C, device="cuda", dtype=dtype)
    w = torch.ones(C, device="cuda", dtype=dtype)
    out_full = F.layer_norm(x, (C,), weight=w, bias=None, eps=1e-5)
    torch.cuda.synchronize()
    bad = 0
    first_bad = -1
    for i in range(0, M, chunk):
        cur = min(chunk, M - i)
        x_c = x[i:i + cur].contiguous()
        ref_c = F.layer_norm(x_c, (C,), weight=w, bias=None, eps=1e-5)
        diff_c = (out_full[i:i + cur].float() - ref_c.float()).abs()
        row_max = diff_c.amax(dim=1)
        nan_row = torch.isnan(row_max)
        bad_mask = (row_max > atol) | nan_row
        n = int(bad_mask.sum().item())
        if n:
            if first_bad < 0:
                first_bad = int(bad_mask.nonzero().flatten()[0].item()) + i
            bad += n
    return {"M": M, "C": C, "bad_rows": bad, "first_bad": first_bad}


def fmt(M, C, label=""):
    numel = M * C
    numel_log2 = int(math.log(numel, 2))
    print(f"  {label:<28s}  numel = {numel:>13,d} (2^{numel_log2:.1f})  M = {M:>9,d}  C = {C:>4d}  ", end="", flush=True)
    try:
        r = check(M, C)
    except Exception as e:  # noqa: BLE001
        print(f"\033[33mERROR\033[0m  {type(e).__name__}: {str(e)[:60]}")
        return
    status = "BAD" if r["bad_rows"] else "ok "
    color = "\033[31m" if r["bad_rows"] else "\033[32m"
    print(f"{color}{status}\033[0m  "
          f"bad_rows = {r['bad_rows']:>10,d}/{M:>10,d}  "
          f"first_bad = {r['first_bad']:>12,d}  ")


def main():
    print(f"PyTorch: {torch.__version__}    HIP: {torch.version.hip}")
    print(f"Device:  {torch.cuda.get_device_name(0)}")
    print()

    # ─────────────────────────────────────────────────────────────────────
    # Section 1 — real-world case: ~27.5 M rows × 267 cols.
    # The kernel runs to completion but silently corrupts most rows.
    # ─────────────────────────────────────────────────────────────────────
    print("=" * 72)
    print("1. Real-world failing shape")
    print("=" * 72)
    print("  The kernel runs to completion but silently corrupts most rows.\n")
    fmt(5247 * 5247, 267, "  real shape  ")
    print()

    # ─────────────────────────────────────────────────────────────────────
    # Section 2 — the precision boundary at M = 2^23.
    # The kernel raises `invalid configuration argument` at exactly
    # M = 2^23; for M > 2^23 it runs but produces silently wrong output.
    # ─────────────────────────────────────────────────────────────────────
    print("=" * 72)
    print("2. Precision boundary at M = 2^23 = 8,388,608")
    print("=" * 72)
    print("  largest OK / at boundary (kernel errors) / smallest BAD,")
    print("  using a single C (=257) to keep it short:\n")
    fmt(2**23 - 1, 257, "  largest OK")
    fmt(2**23,     257, "  at boundary (errors)")
    fmt(2**23 + 1, 257, "  smallest BAD")
    print()

    # ─────────────────────────────────────────────────────────────────────
    # Section 3 — at M = 2^23 + 1, the failure is determined by C % 4.
    # Any C with C % 4 != 0 fails.
    # Any C with C % 4 == 0 is fine.
    # ─────────────────────────────────────────────────────────────────────
    print("=" * 72)
    print("3. At M = 2^23 + 1, the C predicate is `C % 4 != 0`")
    print("=" * 72)
    print("  not divisible by 4: BAD\n")
    for C in [2, 3, 5, 6, 10, 14, 65, 129, 257, 267]:
        fmt(2**23 + 1, C, f"  C = {C}  (C % 4 = {C % 4})")
    print("\n  divisible by 4: ok\n")
    for C in [4, 8, 12, 16, 24, 128, 256, 384, 1024]:
        fmt(2**23 + 1, C, f"  C = {C}  (C % 4 = 0)")
    print()

    # ─────────────────────────────────────────────────────────────────────
    # Section 4 — M is the trigger, NOT numel. Sizes chosen to fit in
    # MI300X's 192 GB with comfortable headroom (peak ~50 GB per case).
    # All three use C % 4 != 0, so the only thing varied is the M/numel
    # combination.
    #  - M = 1, large C       (numel = 5 G):    ok  (numel >> 25 M)
    #  - M = 2^22, large C    (numel = 8.6 G):  ok  (M < 2^23)
    #  - M = 2^23 + 1, tiny C (numel = 25 M):   BAD (M > 2^23 trips it)
    # ─────────────────────────────────────────────────────────────────────
    print("=" * 72)
    print("4. M is the trigger, NOT numel")
    print("=" * 72)
    print("  Tiny M, huge numel:")
    fmt(1, 5_000_000_001, "    M=1, C=5 G+1")          # numel 5.0 G, ~30 GB peak
    print("\n  Larger numel, M still under 2^23:")
    fmt(2**22, 2049, "    M=2^22, C=2049")             # numel 8.6 G, ~35 GB peak
    print("\n  Much smaller numel, but M just past 2^23:")
    fmt(2**23 + 1, 3, "    M=2^23+1, C=3")             # numel 25 M, trivial


if __name__ == "__main__":
    main()
```

</details>

## Related

- #136291 — original ROCm `layer_norm` crash on large tensors
- #144007 — ROCm chunked-launch fix for the crash above
- #181555 — CUDA `layer_norm` wrong-output bug for `M*N > 2^32`
- #181600 — `int64_t` cast fix for #181555 (merged; the nightly tested includes it)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

### Versions

<details><summary>For latest nightly on MI300X</summary>

```
Collecting environment information...
PyTorch version: 2.13.0.dev20260521+rocm7.2
Is debug build: False
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: 7.2.53211

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jan 26 2026, 14:55:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-1074-oracle-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration: AMD Instinct MI300X (gfx942:sramecc+:xnack-)
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: 7.2.53211
MIOpen runtime version: 3.5.1
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               224
On-line CPU(s) list:                  0-223
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8480+
CPU family:                           6
Model:                                143
Thread(s) per core:                   2
Core(s) per socket:                   56
Socket(s):                            2
Stepping:                             8
CPU max MHz:                          3800.0000
CPU min MHz:                          800.0000
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx
fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology
nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid
dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb
cat_l3 cat_l2 cdp_l3 invpcid_single intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid
ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb
intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local
split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku
ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect
cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8
flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            5.3 MiB (112 instances)
L1i cache:                            3.5 MiB (112 instances)
L2 cache:                             224 MiB (112 instances)
L3 cache:                             210 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-55,112-167
NUMA node1 CPU(s):                    56-111,168-223
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Vulnerable; BHI: Vulnerable
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] torch==2.13.0.dev20260521+rocm7.2
[pip3] triton-rocm==3.7.0+git88b227e2
[conda] Could not collect
```

</details>

### PR DESCRIPTION
**Pull request by naromero77amd**
Fixes https://github.com/pytorch/pytorch/issues/184826

## Summary
- Bound the ROCm non-vectorized `layer_norm` fallback launch so `gridDim.x * blockDim.x` stays within HIP's launch limit.
- Use an in-kernel grid-stride loop over rows for the ROCm fallback, following the same style as prior ROCm launch-limit fixes in `roll` and `triu`/`tril`.
- Keep the existing CUDA fallback launch path unchanged.
- Preserve existing vectorized path behavior while covering `C % 4 != 0` cases that skip vectorization.
- Add a regression unit test for the large-`M`, non-vectorized `layer_norm` path.

## Alternatives considered
I first implemented host-side chunking, mirroring the existing vectorized `layer_norm` workaround. That fixed correctness, but it can require multiple rowwise and forward launches for large `M`.

I also evaluated the approach used by https://github.com/pytorch/pytorch/pull/169474 and https://github.com/pytorch/pytorch/pull/179717: bound the ROCm grid and let the kernel walk remaining work with a grid-stride loop. That also fixes correctness while preserving two fallback launches total.

On MI300X / ROCm 7.2, the grid-stride approach was comparable near the boundary and faster on the real failing shape:

| shape | M | N | host chunking median ms | grid-stride median ms |
| --- | ---: | ---: | ---: | ---: |
| boundary_n3 | 8,388,608 | 3 | 27.163 | 27.185 |
| over_boundary_n3 | 8,388,609 | 3 | 27.159 | 27.187 |
| over_boundary_n257 | 8,388,609 | 257 | 33.883 | 34.483 |
| over_boundary_n267 | 8,388,609 | 267 | 34.417 | 34.722 |
| real_n267 | 27,531,009 | 267 | 112.559 | 108.288 |
| real_n268_vectorized | 27,531,009 | 268 | 40.192 | 40.263 |

Given the real repro shape improves and the near-boundary cases are close, this PR uses the grid-stride row loop.

## Test plan
- Added `TestNN.test_layer_norm_large_m_non_vectorized` in `test/test_nn.py` to cover `M=2^23`, `M=2^23+1`, and `N=3` (`N % 4 != 0`).
- `python test/test_nn.py TestNN.test_layer_norm_large_m_non_vectorized`
- Verified previously failing cases now pass with `bad_rows=0`, including `M=2^23`, `M=2^23+1`, `C % 4 != 0`, and the real shape `M=5247*5247, C=267` on MI300X / ROCm 7.2.

Made with [Cursor](https://cursor.com)

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @pragupta @jerrymannil @xinyazhang @manuelcandales @angelayi

### CODE REVIEW

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 1]
```diff
@@ -1156,13 +1174,28 @@ void LayerNormKernelImplInternal(
     launch_vectorized_layer_norm_kernel<T, T_ACC, rms_norm>(static_cast<int>(N), M, eps, X_data, gamma_data, beta_data, Y_data, mean_data, rstd_data);
   } else {
   cudaStream_t cuda_stream = at::cuda::getCurrentCUDAStream();
+#ifdef USE_ROCM
+  // ROCm rejects launches where gridDim.x * blockDim.x exceeds uint32_t max.
+  // Bound the fallback launch and let each block stride over remaining rows.
+  constexpr int64_t max_rowwise_blocks =
+      std::numeric_limits<uint32_t>::max() / cuda_utils::kCUDABlockReduceNumThreads;
+  const dim3 blocks(static_cast<uint32_t>(std::min(M, max_rowwise_blocks)));
```
Comment (jerrymannil) [inline aten/src/ATen/native/cuda/layer_norm_kernel.cu:1182]: Can you also try with different block config ?
We tend to get better perf with smaller grids.

Also good to do a benchmark with different input sizes before and after this change. (small, medium, large) 
```
num_mp = at::cuda::getCurrentDeviceProperties()->multiProcessorCount;
block = (num_mp * X)   // X:  4 - 64
```

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu [hunk 2]
```diff
@@ -1156,13 +1174,28 @@ void LayerNormKernelImplInternal(
...
       <<<M, cuda_utils::kCUDABlockReduceNumThreads, 0, cuda_stream>>>(
-          N, eps, X_data, mean_data, rstd_data);
+          M, N, eps, X_data, mean_data, rstd_data);
   C10_CUDA_KERNEL_LAUNCH_CHECK();
   LayerNormForwardCUDAKernel<T, T_ACC, rms_norm><<<M, kCUDANumThreads, 0, cuda_stream>>>(
-      N, X_data, mean_data, rstd_data, gamma_data, beta_data, Y_data);
+      M, N, X_data, mean_data, rstd_data, gamma_data, beta_data, Y_data);
   C10_CUDA_KERNEL_LAUNCH_CHECK();
+#endif
   }
```
Comment (jerrymannil) [inline aten/src/ATen/native/cuda/layer_norm_kernel.cu:1199]: You can also try to change the vectorized path in a follow-up PR

File: aten/src/ATen/native/cuda/layer_norm_kernel.cu (hunk 1 above)
Comment (dnikolaev-amd) [reply aten/src/ATen/native/cuda/layer_norm_kernel.cu:1182]: > We tend to get better perf with smaller grids

We need to move the loop inside the kernel in in this case

**Review by jeffdaily — APPROVED** [review]
Approve.  CI failures are not related.  Any follow-up suggestions can be a separate PR.
